src/approach/gdumb.py

```diff
@@ -0,0 +1,107 @@
+        return ExemplarsDataset
+
+    @staticmethod
+    def extra_parser(args):
+        """Returns a parser containing the approach specific parameters"""
+        parser = ArgumentParser()
+        parser.add_argument('--regularization', default='cutmix', required=False,
+                            help='Use regularization (default=%(default)s)')
+        return parser.parse_known_args(args)
+
+    def train_loop(self, t, trn_loader, val_loader):
+        """Contains the epochs loop"""
+        # 1. GDumb resets the network before learning a new task, relying only on the exemplars stored so far
+        if t == 0:
+            # Keep the randomly initialized model from time step 0
+            self.init_model = deepcopy(self.model)
+        else: 
+            # Reinitialize the model (backbone) for very task from time step 1
+            self.model.model = deepcopy(self.init_model.model)
+            for layer in self.model.heads.children():
+                layer.reset_parameters()
+
+        # EXEMPLAR MANAGEMENT -- select training subset from current task and exemplar memory
+        aux_loader = torch.utils.data.DataLoader(trn_loader.dataset + self.exemplars_dataset,
+                                                 batch_size=trn_loader.batch_size,
+                                                 shuffle=True,
+                                                 num_workers=trn_loader.num_workers,
+                                                 pin_memory=trn_loader.pin_memory)
+        self.exemplars_dataset.collect_exemplars(self.model, aux_loader, val_loader.dataset.transform)
+
+        # Set new set of exemplars as the only data to train on
+        trn_loader = torch.utils.data.DataLoader(self.exemplars_dataset,
+                                                 batch_size=trn_loader.batch_size,
+                                                 shuffle=True,
+                                                 num_workers=trn_loader.num_workers,
+                                                 pin_memory=trn_loader.pin_memory)
+
+        # FINETUNING TRAINING -- contains the epochs loop
+        super().train_loop(t, trn_loader, val_loader)
+
+    def train_epoch(self, t, trn_loader):
+        """Runs a single epoch"""
+        self.model.train()
+        if self.fix_bn and t > 0:
+            self.model.freeze_bn()
+        for images, targets in trn_loader:
+            # Get exemplars
+            if len(self.exemplars_dataset) > 0:
+                # 2. Balanced batches
+                exemplar_indices = torch.randperm(len(self.exemplars_dataset))[:trn_loader.batch_size]
+                images_exemplars, targets_exemplars = default_collate([self.exemplars_dataset[i]
+                                                                       for i in exemplar_indices])
+                images = torch.cat((images, images_exemplars), dim=0)
+                targets = torch.cat((targets, targets_exemplars), dim=0) 
```

Comment:
@mmasana if the trn_loader is overridden with the exemplar_dataset, why here we do the permutation and not simply use images/targets out of it? It's more like drawing samples with the repetition than balanced batches, right?